crates/wasm/src/storage.rs

```diff
@@ -0,0 +1,179 @@
+                &base64::engine::general_purpose::STANDARD,
+                id.to_proto().inner,
+            ))?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
+    }
+
+    pub async fn get_note(
+        &self,
+        commitment: &note::StateCommitment,
+    ) -> WasmResult<Option<SpendableNoteRecord>> {
+        let tx = self.db.transaction_on_one("spendable_notes")?;
+        let store = tx.object_store("spendable_notes")?;
+
+        store
+            .get_owned(base64::Engine::encode(
+                &base64::engine::general_purpose::STANDARD,
+                commitment.to_proto().inner,
+            ))?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
+    }
+
+    pub async fn get_note_by_nullifier(
+        &self,
+        nullifier: &Nullifier,
+    ) -> WasmResult<Option<SpendableNoteRecord>> {
+        let tx = self.db.transaction_on_one("spendable_notes")?;
+        let store = tx.object_store("spendable_notes")?;
+
+        store
+            .index("nullifier")?
+            .get_owned(&base64::Engine::encode(
+                &base64::engine::general_purpose::STANDARD,
+                nullifier.to_proto().inner,
+            ))?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
+    }
+
+    pub async fn store_advice(&self, note: Note) -> WasmResult<()> {
+        let tx = self.db.transaction_on_one("notes")?;
+        let store = tx.object_store("notes")?;
+
+        let note_proto: penumbra_proto::core::crypto::v1alpha1::Note = note.clone().try_into()?;
+        let note_js = serde_wasm_bindgen::to_value(&note_proto)?;
+
+        let commitment_proto = note.commit().to_proto();
+
+        let commitment_js = serde_wasm_bindgen::to_value(&commitment_proto)?;
+
+        store.put_key_val_owned(commitment_js, &note_js)?;
+
+        Ok(())
+    }
+
+    pub async fn read_advice(&self, commitment: note::StateCommitment) -> WasmResult<Option<Note>> {
+        let tx = self.db.transaction_on_one("notes")?;
+        let store = tx.object_store("notes")?;
+
+        let commitment_proto = commitment.to_proto();
+
+        let commitment_js = serde_wasm_bindgen::to_value(&commitment_proto)?;
+
+        store
+            .get_owned(commitment_js)?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
+    }
+
+    pub async fn get_chain_parameters(&self) -> WasmResult<Option<ChainParameters>> {
+        let tx = self.db.transaction_on_one("chain_parameters")?;
+        let store = tx.object_store("chain_parameters")?;
+
+        store
+            .get_owned("chain_parameters")?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
```

Comment:
I would not convert this to an Option as every consumer is treating it as a Result.

Reply:
Same feedback for the rest of these as well

Reply:
I proceeded from the fact that the situation when we successfully read data from indexedDb by a given key, but there was no record by this key in indexedDB, should not lead to an error.

Reply:
It feels easier to make this a Result as we are nearly always expecting data to be there. It looks like there is only one Option case that could easily work the same as a Result. If you feel strongly otherwise, I can live with it though.

Reply:
Another point is that when initially reading from indexedDB, we get `Option<JsValue> `
```rust
 let read_result: Option<JsValue> = store
            .get_owned("chain_parameters")?
            .await?;
``` 
This means we'd need to convert `Option -> Result`,  and it doesn't seem to be very convenient
For some functions like `get_chain_parameters()` and `get_fmd_parameters()`, we do expect that either the record exists or something is wrong. 
But in other cases, there is a high probability that we will successfully read from indexedDb, but the data will not be there. And in the future, if we want more interaction with indexedDb, there will be more logic similar to 
```rust
  let option = storage.get_note_by_nullifier().await?;
    match option {
        None => {
            // forget nullifier
        }
        Some(note_record) => {
            // Mark record_note as spent
        }
    }
```